### ISSUE TITLE
Does gcc6 produce ppc64 binaries? Macports gcc6 fails

### ISSUE THREAD

**Issue by barracuda156** [iains/gcc-6-branch#1]
@Iains Could you please confirm if your gcc6 / gcc7 produce working ppc64 binaries? With Macports versions it is broken: https://trac.macports.org/ticket/64260

**Comment by iains** [comment]
It seems you trying to build -m64 with a powerpc-darwin9 ...

- all my branches definitely do the first (I always test {-m32,-m64}) ..

```
This GDB was configured as "powerpc-apple-darwin".
(gdb) exec-file /Users/svacchanda/Dev/HelloWorld 
unable to read unknown load command 0x24
unable to read unknown load command 0x26
```
unknown load commands suggests that you are using a later version of cctools than the one from Xcode 3.1.4.

NOTE1: some later versions of ```cctools as``` are broken for ppc64 (on darwin9/macOS 10.5 [I usually use the assembler and linker installed by Xcode 3.1.4]).

(However, I do have some patches to fix ```cctools as``` for ppc64, if you have not already applied them - e.g. https://github.com/iains/darwin-xtools/commit/c8f66d9eb4d1afc3748af7f60fa0c9ce16aa6220 - if you let me know which version of cctools you are using I might be able to point you at specific ppc64 fixes).

NOTE2: for larger binaries, there are some problems with branch islands for ld64-85.2.1. 
( I think that macports has some fixes applied - I ended up rewriting the branch islanding code but that re-write is only on later versions of ld64. )

Unless  you need Java, I'd suggest using a later compiler - gcc-7.5 at least (but there are some pending critical patches to be applied to my github branch there - I've been busy with the arm64 port so older branches are a bit behind on back-ports for fixes).

**Comment by barracuda156** [comment]
@iains Thank you for replying! Initially I bumped into the problem that icu fails to build as universal (ppc+ppc64), then @kencu suggested to check if compiler actually builds ppc64 binaries – and no, both gcc6 and gcc7 apparently fail.
Even Helloworld built by gcc7 as 64-bit fails, while the same built with Apple gcc works:

```
36-225:Dev svacchanda$ gcc-mp-7 -arch ppc HelloWorld.c -o HelloWorld
HelloWorld.c:2:1: warning: return type defaults to 'int' [-Wimplicit-int]
 main() {
 ^~~~
36-225:Dev svacchanda$ /Users/svacchanda/Dev/HelloWorld 
hello, world
36-225:Dev svacchanda$ gcc-mp-7 -arch ppc64 HelloWorld.c -o HelloWorld
HelloWorld.c:2:1: warning: return type defaults to 'int' [-Wimplicit-int]
 main() {
 ^~~~
36-225:Dev svacchanda$ /Users/svacchanda/Dev/HelloWorld 
Segmentation fault
36-225:Dev svacchanda$ file /Users/svacchanda/Dev/HelloWorld 
/Users/svacchanda/Dev/HelloWorld: Mach-O 64-bit executable ppc64
36-225:Dev svacchanda$ gcc -arch ppc64 HelloWorld.c -o HelloWorld
36-225:Dev svacchanda$ /Users/svacchanda/Dev/HelloWorld 
hello, world
36-225:Dev svacchanda$ file /Users/svacchanda/Dev/HelloWorld 
/Users/svacchanda/Dev/HelloWorld: Mach-O 64-bit executable ppc64
```

(I do have both gcc6 and gcc7 built with Macports and working for 32-bit binaries, both on 10.5.8 and 10.6 PPC. But I try to build up the whole ecosystem on 10.5.8 as universal – and here number of things are broken atm.)

Another major problem is that llvm-3.3 and llvm-3.4 fail to build as universal: https://trac.macports.org/ticket/64253
But ld64 wants them as universal.

Finally, libquadmath does not build on PPC even as 32-bit. I opened a ticket for it too: https://trac.macports.org/ticket/64296

I will be grateful for any suggestions.

Which compiler would you recommend me to start with? I can work outside of Macports. If I get a compiler that works for ppc+ppc64 correctly, then it is apparently Macports that does something wrong.

For now, I want to build everything in 10.5.8 PPC and hopefully as ppc+ppc64. (Once that is done, I will replicate the process for 10.6 PPC (10A190), but perhaps for ppc only.)

**Comment by iains** [comment]
> I will be grateful for any suggestions.

Please leave it with me for a little while; I would like to check for the critical back ports and push those to github for you.

> Which compiler would you recommend me to start with?

It depends on your objectives:
 * if you want Java, then you are constrained to 6.5 max (and that probably needs some backports to be effective for building things like LLVM)
 * if you want a capable C++11 compiler useful for building m32/m64 projects, then I would suggest using something later - at least 7.5 (there were some important fixes made for building more complex C++ codes).  

> I can work outside of Macports.

At least at first it would be useful to check that you have a working baseline.

> For now, I want to build everything in 10.5.8 PPC and hopefully as ppc+ppc64.

As noted, I *always* build GCC with multilibs enabled and test m64, so I know that this is possible;


>> (Once that is done, I will replicate the process for 10.6 PPC (10A190), but perhaps for ppc only.)

You cannot build ppc64 on X86 Rosetta [Rosetta only supports ppc7400 and there are no ppc64 slices in the X86 10.6 install].

- can you give me a pointer to 10A190 and where that would be available?

**Comment by barracuda156** [comment]
@iains I am not concerned with Java, at least as of now, and ideally would like to have gcc11 built. However anything from gcc8 onward have been failing so far, both in Macports and outside. (I know that gcc8 must build, since several people have done that, but perhaps I lack some components to make it work.)

Since building any gcc is a multi-hour process on G5, and configuration may fail, and then I need to start over, it would be really helpful to have an idea how to do it in a way you have tested on 10.5.8 PPC.
Once I have baseline system fully functional, I can experiment.

Surely, I can wait until you implement current fixes.

Links to several builds of 10.6 PPC are available in the wikipost here: https://forums.macrumors.com/threads/snow-leopard-on-unsupported-ppc-machines.2232031/
I have 10A190 running on G5, it is fully functional (minor bugs are there): https://forums.macrumors.com/threads/snow-leopard-on-unsupported-ppc-machines.2232031/page-8?post=28428563#post-28428563 (download link in this comment).
So it is not Rosetta, it is natively-run 10.6 on PowerPC. Xcode 3.2 that I use is from 10A190 build, though I have tested newer Xcode and Apple gcc from 10A222, they work too, but require replacing ld and gnumake with those from 10A190. There were later builds functional on PowerPC, but sadly we cannot find those.

I have ran into these notes recently: https://opensource.apple.com/source/dyld/dyld-239.3/doc/ReleaseNotes.txt.auto.html
Between 10A190 which fully works on PPC and 10A222 which partly boots there were several builds. Interestingly, support for PPC for dyld is said to be removed after 10.7 (!).

**Comment by iains** [comment]
> @iains I am not concerned with Java, at least as of now,

OK,  let's work on 10.3+some backports which I am sorting out because ...

> and ideally would like to have gcc11 built. 

... gcc11+ needs a C++11 compiler to bootstrap, so that gcc10 is the last one that can be bootstrapped with Apple gcc-4.2.1 (and I will make some effort to ensure that we back port fixes to 10.x and keep it bootstrappable with gcc.4.2.1)


> Since building any gcc is a multi-hour process on G5,

around 4 hrs on my QuadG5, so yes it's better not to start over.

> Surely, I can wait until you implement current fixes.

GCC-10.3 release builds and works fine on powerpc-darwin9 (including ppc64):

[summ-x86_64-darwin10-r10-9664-10-3-gcc-421.txt](https://github.com/iains/gcc-6-branch/files/7776974/summ-x86_64-darwin10-r10-9664-10-3-gcc-421.txt)

**but** I do suggest to wait for a few backports (that will be in 10.4)

Once this is done _ i will post a gcc-10-3-ppc branch to my gcc-10 repo and add a note here.

> Links to several builds of 10.6 PPC are available in the wikipost here: 

Thanks, I will probably investigate, although such "unofficial/never-officially-released" resources are always a bit tricky to rely on.

**Comment by iains** [comment]
> Once this is done _ i will post a gcc-10-3-ppc branch to my gcc-10 repo and add a note here.

done - please move ay extra discussion to that repo.

transcript of the build on QG5 (using Apple gcc-4.2.1 as  the bootstrap compiler, so no Ada - I usually bootstrap with GCC 7.5 with Ada - but you don't have that).

===== Works For Me™
```
$ /src-local/gcc-git-10/configure --prefix=/opt/iains/powerpc-apple-darwin9/gcc-10-3-ppc --build=powerpc-apple-darwin9 --with-as=/usr/bin/as --with-ld=/usr/bin/ld --enable-languages=all CC=gcc-4.2 CXX=g++-4.2 >conf.txt
configure: WARNING: using in-tree isl, disabling version check
configure: WARNING: GNAT is required to build ada
configure: WARNING: brig not supported for this target
configure: WARNING: d not supported for this target
configure: WARNING: go not supported for this target
configure: WARNING: --enable-host-shared required to build jit
*** This configuration is not supported in the following subdirectories:
     target-libsanitizer target-libvtv target-libphobos target-libgo gnattools gotools target-libada target-libhsail-rt target-zlib target-libffi target-liboffloadmic
    (Any other directories should still work fine.)
rm: conftest.dSYM: is a directory

$ time nice make -j4 STAGE1_CFLAGS="-O1 -pipe -mdynamic-no-pic" STAGE1_CXXFLAGS="-O1 -pipe -mdynamic-no-pic" >b.txt 2>e.txt && time make install >ins.txt

^^ using O1 makes the stage 1 compiler quicker, which speeds up the bootstrap, **but** one has to take care that this does not demand too much of the older compiler used for bootstraps.

well, not so long to build actually ...
real	126m36.050s
user	302m21.635s
sys	60m45.401s

real	0m55.869s
user	0m20.198s
sys	0m25.986s

$ /opt/iains/powerpc-apple-darwin9/gcc-10-3-ppc/bin/gcc /source/test/hello.c -m64 -o hello

$ ./hello
C Hello (count = 0) from 10.3.0

$ file hello
hello: Mach-O 64-bit executable ppc64
```

**Comment by iains** [comment]
FAOD you want this branch:
https://github.com/iains/gcc-10-branch/tree/gcc-10-3-ppc

That has a few recent patches back-ported to 10.3 (those patches will be part of 10.4, when that is released - but not due for some time)

**Comment by kencu** [comment]
Iain, do we know the exact versions of the supporting libraries you are using too, just for 100% reproducibility?

thanks for doing all this, @barracuda156 has been very keen on MacOS, darwin, PowerPC, esp ppc64, and he has hopes of making 10.6 for PPC more than it currently is -- nice to help him along!

All the best this Christmas season, hope all is well in the Jolly Old homeland (we're apparently lowland Scots -> plantation Northern Irish -> Canadian as of 1806 ish).

**Comment by iains** [comment]
For toolchain builds, I remove all external dependencies (otherwise the permutations become unmanageable).

So GMP, MPFR, MPC and ISL are bootstrapped along with the compiler and the only external deps are in /usr/lib [ I just symlink the sources into the GCC source tree - it's a very quick process]

For the record the versions of GMP et al I used are (as reported by -v):
```
GMP version 6.1.2, MPFR version 3.1.6, MPC version 1.1.0, isl version isl-0.20-GMP
```
but I would expect it to work with any supported editions.

**Comment by iains** [comment]
and indeed Seasons' Greetings all round :) ...

..  I've been doing a little tinkering (but TBH most effort at the moment is going into the M1/Arm64 port - ppc and x86 are pretty much a 'done deal' - of course, with plenty of bugs still to fix ... )

**Comment by kencu** [comment]
Has anyone put together some kind of script that allows you to install a prebuilt gcc into a prefix of your choosing, and adjusts what is needed to make it find the required parts?

Seems like if we had such an animal that people could download and install to bootstrap from for PPC, then lots of folks would be miles ahead.

And if not that, perhaps an installer that would go into a location that nobody else would ever use (/opt/gcc/gccN ?) or something would be second best.

This bootstrapping gcc on older systems and ppc in particular does drive some people to drink...

**Comment by iains** [comment]
There needs to be one more change to the GCC tree to allow the installed compiler to be "relocatable" (which actually is the ```@rpath``` one) - I posted that change for review a few weeks ago, and hope to get that into GCC12.

(then, of course, it could be backported or used to make a compiler as you suggest).

FWIW; I also build native-crosses (so a powerpc-darwin9 compiler built on x8_64-darwin20, for example) - but such things are not beginner's projects - it would need to be scripted or something.

**Comment by iains** [comment]
> And if not that, perhaps an installer that would go into a location that nobody else would ever use (/opt/gcc/gccN ?) or something would be second best.


Actually, that would not be too hard to do - I did something along those lines for GCC-5 .. which installed into /opt/iains/xxxxxxx/ for the reasons  you describe.  I guess the main issue if we wanted to do this for every live macOS version would be somewhere to host it - not sure if github would allow me that much space for a release.

The neat thing there would be that it could include Ada (and eventually D) which are both needed to bootstrap those languages.

> This bootstrapping gcc on older systems and ppc in particular does drive some people to drink...

Apart from the slowness of the older machines, there's nothing special about the powerpc-darwin port - it's being working pretty well since gcc-5-ish era (there was a time between 4.2.1 and then were there was not much active maintainance - but, TBH, that also affected X86).

Anyway - the changes in that 10.3 branch fix the libgcc_s problems (we still have to find a good solution to libstdc++ .. )

**Comment by kencu** [comment]
People seem to host WAY more than that on github. There is a guy who builds a 500MB Chromium every 12 hours and hosts that on github :> So go for it, I would say!

You can expect lots of ongoing pushback about rpath from the MacPorts admins, as it causes them heartburn. I hear about it every week.  If gcc used rpath internally, but built software with full library paths by default, would that do it for you? That is exactly what clang/llvm do on MacPorts, and people fully accept that.

**Comment by iains** [comment]
> You can expect lots of ongoing pushback about rpath from the MacPorts admins, as it causes them heartburn. I hear about it every week. If gcc used rpath internally, but built software with full library paths by default, would that do it for you? That is exactly what clang/llvm do on MacPorts, and people fully accept that.

It cannot work on macOS 10.11+ (i.e. on all systems with SIP including the new arm64 port) - this is because the effects of SIP disable DYLD_LIBRARY_PATH through key executables - so that the GCC configuration and build does not work properly.  For the earlier systems ( < 10.11), fixed install paths can work as they always did before.

However, even on those [earlier] systems, for the compiler to be relocatable, the runtime libraries must be built as @rpath - since otherwise it can only be installed into the fixed position it was configured for (which is undesirable for some people because they might only have write permissions for their home directory - so an /opt/xxxx is probably no use to them  - although perhaps /Users/Share/xxxx would be OK).

TBH the end user should not see any difference if the compiler is installed into /opt/macports.... or /Users/Shared . or ~/gcc-n

FWIW re LLVM, the normal build uses @rpath for the runtimes - so you must have some local change to that:

this is yesterday's LLVM main:

```
$ otool -Lv ./lib/libc++.dylib 
./lib/libc++.dylib:
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
	@rpath/libc++abi.1.dylib (compatibility version 1.0.0, current version 1.0.0)

$ otool -Lv ./lib/libc++abi.dylib 
./lib/libc++abi.dylib:
	@rpath/libc++abi.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

**I am really not on a crusade here - folks are free to use stuff or not as they wish - my point remains that @rpath is Apple's design to solve relocation problems, so I am not sure why we would be concerned to use it.  The clincher for me is that I do not see a way to build toolchains reliably on systems with SIP unless we make use of it.**

**Comment by kencu** [comment]
It's a MacPorts thing, based on their super-strict isolation policy.

```/opt/local/lib```is not on any default rpath search path, so everything they build has to have that specifically added to every library and executable as an added search path that comes before all the standard search paths. And that causes lots more trouble than you would think it should :>, given such a plethora of software build systems out there.

If we dumped everything in ```/usr/local/``` like other package managers do, our lives would be 1000 times easier, as that location is on every compiler's default search path and nothing needs to be done to find it.

**Comment by kencu** [comment]
I set up the build of llvm / clang to use rpath references internally (for it's own tools to find the right libraries) but when clang/llvm builds other software, by default that software uses full library path reference.

People have accepted that. Our installed newer-than-the-system-libc++  libc++.dylib has a full path reference in it.

**Comment by iains** [comment]
> It's a MacPorts thing, based on their super-strict isolation policy.
> 
> `/opt/local/lib`is not on any default rpath search path, so everything we build has to have that specifically added to every library and executable as an added search path that comes before all the standard search paths. And that causes lots more trouble than you would think it should :>, given such a plethora of software build systems out there.

Nothing should change with the compiler's use of ```@rpath```.

Now you have   libfoo.dylib  ... named as /opt/local/lib/libfoo.dylib
... it's still conceptually a runpath /opt/local/lib and a name libfoo.dylib

With libraries which include their runpath, you do not need a runpath in the exe (it's encoded in the library load command effectively)

When using ```@rpath``` you still have the two components - but the run path is supplied by a different load command in the exe.

So compiler (now)
```/opt/local/lib/gcc/xxxxxx/libgfortran.dylib named /opt/local/lib/gcc/xxxxxx/libgfortran.dylib ```
exes have no secondary run path, the installation is fixed.

Compiler with ```@rpath``` changes

```/opt/local/lib/gcc/xxxxxx/libgfortran.dylib is named @rpath/libgfortran.dylib```

exes contain the run path ```/opt/local/lib/gcc/xxxxxx/``` which ia automatically added by the compiler - there is still no change to the default search paths etc.

> If we dumped everything in `/usr/local/` like other package managers do,

Please don't :) I think it's essential for 'distros' to be isolated from each other and from Apple's stuff - and TBH 'default search paths' are not actually always that friendly.

**Comment by kencu** [comment]
I believe MacPorts is going to have to bite the bullet and eventually sort this out -- and stop carrying any of. the older software they like to carry that can't be made to work with rpaths. So do what you need to do !

**Comment by iains** [comment]
> I set up the build of llvm / clang to use rpath references internally (for it's own tools to find the right libraries) but when clang/llvm builds other software, by default that software uses full library path reference.

I think there's a misconception about what the ```@rpath``` compiler change does.
It **_only_** affect the compiler's runtimes and the run paths for them - is does not magically alter other libraries in /opt/local/lib (although they will refer to the compiler's runtimes with two components - run path and name).

GCC with or without the @rpath change still refers to (other, not gcc runtime) libraries with a "full path" in the same way as before.

I guess the other interesting thing is that the arm64 prototype has had this since day #1 - and after a few early bugs were ironed out of some of the packages, there does not seem to be a rash of problems reported.

> People have accepted that. Our installed newer-than-the-system-libc++ libc++.dylib has a full path reference in it. 

So you post-process it or modified the LLVM build process. 

I guess you can always post-process the GCC runtimes to undo the ```@rpath``` and patch the compiler to not insert the run paths.  install_name_tool would work for the libs - but I'm not quite sure how to make the compiler specs work differently after installation (will think some more)

Right now I'm struggling to see what problem is being solved by the fear of rpaths.

**Comment by kencu** [comment]
https://trac.macports.org/ticket/63148

https://trac.macports.org/ticket/64211

https://trac.macports.org/ticket/64204

https://trac.macports.org/ticket/63899

https://trac.macports.org/ticket/63814

https://github.com/macports/macports-ports/pull/13156#issuecomment-989005400

https://trac.macports.org/ticket/64089

**Comment by kencu** [comment]
I could find many more, I think. Lots have been closed as duplicates of others. The MacPorts admins vehemently dissuade rpath usage at all costs, at present.

But it has to get sorted out...

**Comment by iains** [comment]
I am open to alternate technical solutions ; :D
someone could rewrite auto-tools, the GCC build system and dejagnu - but that seems rather unlikely - that would be a mammoth project.

Some things can be converted to alternate build systems - I've occasionally switched auto* based builds to cmake - but that is not going to work for the complex case of GCC's build (some build combinations that GCC does easily are actually pretty challenging for other build systems).

I do agree that folks have and do misunderstand how to use rpaths, but that's just another source of bugs - I do not see it's any worse than misunderstanding any other aspect of programming.  The intent of the GCC change is to be transparent to the end user.

**Comment by kencu** [comment]
@ryandesign : I know you've hoping for an opportunity to weigh in on the rpath use in gcc with @iains -- here's a good opportunity for you to say your piece if you would like to do so now.

**Comment by kencu** [comment]
What I would be proposing would be something that would use rpaths during the building and testing of gcc, but on installation, would rewrite these rpaths to be full paths.

That is what meson and cmake do by default at present, and that works pretty well. 

I'm not certain what kind of tooling trouble that might cause -- I certainly can see that you are keeping gcc afloat as more-or-less a one man show, and whatever makes your life easier and able to keep that going is much valued.

But if what I am describing seems pretty simple, or at least simple for us to do with a post-install script, then I know that at present at least, that would go over more easily with the MacPorts admins, who do see troubles with gcc at present that seem related to rpath use and seem hard to sort out.

**Comment by ryandesign** [comment]
Ken has already provided several examples of rpath-induced problems in MacPorts. Here's another that I resolved by disabling the use of rpath:

https://trac.macports.org/ticket/61735

I myself know very little about what rpath is or does, and it doesn't particularly interest me. rpath did not exist on macOS prior to 10.5. Somehow we got by without it until then, therefore my view is that it is unnecessary and should be avoided. MacPorts still runs on 10.4 and requiring rpath would be an impediment to getting a port working on 10.4. (Assuredly, in many cases, there would be many more obstacles than that.)

DYLD_LIBRARY_PATH works fine with SIP, DYLD_ variables just aren't inherited by subprocesses like they used to be. Set DYLD_LIBRARY_PATH right before you run an executable that needs it. Don't set it for example at the top level of a Makefile and expect it to be passed down to all sub-Makefiles and scripts.

Relocatability is not a concern for MacPorts. Users can still relocate software (i.e. copy MacPorts-installed software elsewhere for use within their own projects) by altering the install names after the fact, either with install_name_tool or dylibbundler. Relocatability could also be accomplished prior to macOS 10.5 without the use of `@rpath` (e.g. `@executable_path` and `@loader_path`).

**Comment by iains** [comment]
>I myself know very little about what rpath is or does, and it doesn't particularly interest me. rpath did not exist on macOS prior to 10.5. Somehow we got by without it until then, therefore my view is that it is unnecessary and should be avoided. 

- well macOS has changed quite a bit between darwin8 and darwin19 .. and even more for aarch64-darwin2* (there are more constraints on what we can do there).

- rpaths are not needed for correctness on <= darwin14 so it makes 0 difference to 10.4 (which built gcc-11 fine for me last I tried).

like I said, you could rewrite auto* to fix the problem

> DYLD_LIBRARY_PATH works fine with SIP, DYLD_ variables just aren't inherited by subprocesses like they used to be. Set DYLD_LIBRARY_PATH right before you run an executable that needs it. Don't set it for example at the top level of a Makefile and expect it to be passed down to all sub-Makefiles and scripts.

* Yes, we all know this ....
.. however, in order to implement what you describe you need either to re-write autoconf & friends (there are hardwired uses of /bin/sh) or use some proxy for the DYLD_LIBRARY_PATH and then pass that past the uses of system binaries (like /bin/sh, env, etc).

* JFTR, I actually implemented a partial scheme for the second idea in GCC - but the problem is that then one needs to implement it for GMP, MPFR, MPC and ISL otherwise in-tree builds of those don't work   in the end it is not a viable piece of work for volunteer time.

* configuration options are reversible, you would always have the option to disable it... and keep the broken behaviour.

* As with all OSS stuff - - if you have a better solution "patches are welcome".

I have not any more to add on this topic - as noted earlier - folks are free to choose to use what they like, or write their own patches.

**Comment by barracuda156** [comment]
> FAOD you want this branch: https://github.com/iains/gcc-10-branch/tree/gcc-10-3-ppc
> 
> That has a few recent patches back-ported to 10.3 (those patches will be part of 10.4, when that is released - but not due for some time)

Thank you very much! I will try building it and update you accordingly.

If that works, next I wanna build your updated darwin xtools.

**Comment by iains** [comment]
> Thank you very much! I will try building it and update you accordingly.

If there are any issues - please report them on the gcc-10 repo :)

 
> If that works, next I wanna build your updated darwin xtools.

For Darwin9/10 era it probably makes more sense to build xtools-1.1.0 (that will avoid needing to built the libtapi) - I see you had some issues with that - but we should be able to resolve those (I have used that toolset for several years on darwin9)

(as an aside) I did try making an install of 10.6-dp-96 onto a partition - but I am not sure it's going to be much use on a G5 - there seem to be very few of the libraries with a ppc64 slice and there's no 64b ppc kernel.  I guess the process of removing ppc64 was already advanced by then.

**Comment by barracuda156** [comment]
@iains 
> FWIW; I also build native-crosses (so a powerpc-darwin9 compiler built on x8_64-darwin20, for example) - but such things are not beginner's projects - it would need to be scripted or something.

I have tried to build a gcc11 cross-compiler earlier on an Intel Mac (to build for PowerPC), as described here: http://maniacsvault.net/articles/powerpccross
However it ran into errors on some later stages:

```
checking how to recognize dependent libraries... pass_all
checking for powerpc-apple-darwin9.8.0-ar... (cached) /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ar
checking for powerpc-apple-darwin9.8.0-strip... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/strip
checking for powerpc-apple-darwin9.8.0-ranlib... (cached) /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ranlib
checking for gawk... awk
checking command to parse /Users/svacchanda/gcc_all/gcc11_build/./gcc/nm output from /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include    object... /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include    -E
/Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include    -E
checking for grep that handles long lines and -e... checking whether -fstack-protector works... yes
checking for CET support... no
checking whether hidden visibility is supported... yes
checking whether symbol versioning is supported... no
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... failed
checking for powerpc-apple-darwin9.8.0-dsymutil... no
checking for dsymutil... dsymutil
configure: WARNING: using cross tools not prefixed with host triplet
checking for powerpc-apple-darwin9.8.0-nmedit... no
checking for nmedit... nmedit
checking for powerpc-apple-darwin9.8.0-lipo... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/lipo
checking for powerpc-apple-darwin9.8.0-otool... powerpc-apple-darwin9.8.0-otool
checking for powerpc-apple-darwin9.8.0-otool64... no
checking for otool64... no
checking for -single_module linker flag... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... no
checking for -exported_symbols_list linker flag... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
make[1]: *** [configure-target-libobjc] Error 1
yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for stdlib.h... yes
checking for sys/stat.h... yes
checking for string.h... yes
checking for stdlib.h... yes
checking for memory.h... yes
checking for string.h... yes
checking for strings.h... yes
checking for memory.h... yes
checking for inttypes.h... yes
checking for strings.h... yes
checking for stdint.h... yes
checking for inttypes.h... yes
checking for unistd.h... yes
checking for stdint.h... yes
checking minix/config.h usability... yes
checking for unistd.h... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking alloca.h usability... yes
checking for makeinfo... makeinfo --split-size=5000000
checking for modern makeinfo... yes
checking alloca.h presence... yes
checking for alloca.h... yes
checking malloc.h usability... no
checking malloc.h presence... no
checking for malloc.h... no
yes
checking paths.h usability... checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include   ... /Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld
checking if the linker (/Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /Users/svacchanda/gcc_all/gcc11_build/./gcc/nm
checking the name lister (/Users/svacchanda/gcc_all/gcc11_build/./gcc/nm) interface... yes
checking paths.h presence... yes
checking for paths.h... yes
checking syslog.h usability... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking for /Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld option to reload object files... -r
checking for powerpc-apple-darwin9.8.0-objdump... powerpc-apple-darwin9.8.0-objdump
checking how to recognize dependent libraries... pass_all
checking for powerpc-apple-darwin9.8.0-ar... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ar
checking for powerpc-apple-darwin9.8.0-strip... (cached) /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/strip
checking for powerpc-apple-darwin9.8.0-ranlib... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ranlib
checking command to parse /Users/svacchanda/gcc_all/gcc11_build/./gcc/nm output from /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include    object... yes
checking syslog.h presence... yes
checking for syslog.h... yes
checking for string.h... (cached) yes
checking for unistd.h... (cached) yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking stdio.h usability... yes
checking stdio.h presence... failed
checking for powerpc-apple-darwin9.8.0-dsymutil... no
checking for dsymutil... dsymutil
configure: WARNING: using cross tools not prefixed with host triplet
checking for powerpc-apple-darwin9.8.0-nmedit... no
checking for nmedit... nmedit
checking for powerpc-apple-darwin9.8.0-lipo... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/lipo
checking for powerpc-apple-darwin9.8.0-otool... powerpc-apple-darwin9.8.0-otool
checking for powerpc-apple-darwin9.8.0-otool64... no
checking for otool64... no
checking for -single_module linker flag... yes
checking for stdio.h... yes
checking limits.h usability... no
checking for -exported_symbols_list linker flag... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
yes
checking limits.h presence... make[1]: *** [configure-target-libquadmath] Error 1
yes
checking for limits.h... yes
checking whether vsnprintf is usable... no
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include   ... /Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld
checking if the linker (/Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /Users/svacchanda/gcc_all/gcc11_build/./gcc/nm
checking the name lister (/Users/svacchanda/gcc_all/gcc11_build/./gcc/nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking for /Users/svacchanda/gcc_all/gcc11_build/./gcc/collect-ld option to reload object files... -r
checking for powerpc-apple-darwin9.8.0-objdump... powerpc-apple-darwin9.8.0-objdump
checking how to recognize dependent libraries... pass_all
checking for powerpc-apple-darwin9.8.0-ar... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ar
checking for powerpc-apple-darwin9.8.0-strip... (cached) /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/strip
checking for powerpc-apple-darwin9.8.0-ranlib... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ranlib
checking command to parse /Users/svacchanda/gcc_all/gcc11_build/./gcc/nm output from /Users/svacchanda/gcc_all/gcc11_build/./gcc/xgcc -B/Users/svacchanda/gcc_all/gcc11_build/./gcc/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/ -B/usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/lib/ -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/include -isystem /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/sys-include    object... failed
checking for powerpc-apple-darwin9.8.0-dsymutil... no
checking for dsymutil... dsymutil
configure: WARNING: using cross tools not prefixed with host triplet
checking for powerpc-apple-darwin9.8.0-nmedit... no
checking for nmedit... nmedit
checking for powerpc-apple-darwin9.8.0-lipo... /usr/local/gcc/ppc/11/powerpc-apple-darwin9.8.0/bin/lipo
checking for powerpc-apple-darwin9.8.0-otool... powerpc-apple-darwin9.8.0-otool
checking for powerpc-apple-darwin9.8.0-otool64... no
checking for otool64... no
checking for -single_module linker flag... no
checking for -exported_symbols_list linker flag... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
make[1]: *** [configure-target-libssp] Error 1
make: *** [all] Error 2
```

[config-2.log](https://github.com/iains/gcc-6-branch/files/7798944/config-2.log)

Building for ppc or ppc64, against 10.5 or 10.6 SDK resulted in the same error ultimately (if using 10.6 SDK and ppc64, few dylib files had to be replaced with ones from 10.5 SDK to solve arch errors, otherwise no difference).

**Comment by barracuda156** [comment]
> For Darwin9/10 era it probably makes more sense to build xtools-1.1.0 (that will avoid needing to built the libtapi) - I see you had some issues with that - but we should be able to resolve those (I have used that toolset for several years on darwin9)

Yes, I would request you to suggest me a correct sequence to build your xtools. I did try, but something is lacking.

> (as an aside) I did try making an install of 10.6-dp-96 onto a partition - but I am not sure it's going to be much use on a G5 - there seem to be very few of the libraries with a ppc64 slice and there's no 64b ppc kernel. I guess the process of removing ppc64 was already advanced by then.

Yes, unfortunately, it seems that ppc64 support is nearly absent in 10.6 PPC, whichever build. There are few evidence suggesting Apple had internal PPC branch for quite a while past 10A190, but if so, we don’t have those builds.

For ppc32 10A190 is functional though. I have installed Xcode Tools from 10A222 few days ago, its compilers are perfectly functional (`gcc-4.2` build 5626, newer than in 10A96, 10A190 or 10.5.8), but `ld` and `gnumake` have to be replaced from 10A190.
(I never used 10A96.)

Your gcc-10.3-ppc – I rebuilt it with --disable-multilib on 10.5.8 without re-installing – did install on 10A190 and seems to work for ppc32 too. At least I successfully got helloworld with it. It did not want to build natively – perhaps requires some amendments, which I am not qualified to make (setting deployment target and SDK to 10.5 didn’t help, so I decided for the time-being to build on 10.5.8 and use that to install on 10.6).

**Comment by barracuda156** [comment]
Just to keep track, referring here to the newly found error with building `icu` with `gcc6`: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104183

Build fails with the following:

```
ld: duplicate symbol icu_67::number::NumberFormatterSettings<icu_67::number::UnlocalizedNumberFormatter>::copyErrorTo(UErrorCode&) const in number_skeletons.o and number_fluent.o
collect2: error: ld returned 1 exit status
make[1]: *** [../lib/libicui18n.67.1.dylib] Error 1
make: *** [all-recursive] Error 2
```

Commands producing the faulty objects:

```
/opt/local/bin/g++-mp-6 -m64 -O2 -DU_ATTRIBUTE_DEPRECATED= -DU_I18N_IMPLEMENTATION -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -I. -I../common -O2 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++11 -fvisibility=hidden -fno-common -c -dynamic -MMD -MT "number_skeletons.d number_skeletons.o number_skeletons.ao" -o number_skeletons.o number_skeletons.cpp

/opt/local/bin/g++-mp-6 -m64 -O2 -DU_ATTRIBUTE_DEPRECATED= -DU_I18N_IMPLEMENTATION -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -I. -I../common -O2 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++11 -fvisibility=hidden -fno-common -c -dynamic -MMD -MT "number_fluent.d number_fluent.o number_fluent.ao" -o number_fluent.o number_fluent.cpp
```

**Comment by iains** [comment]
if you want to try and track down the changes that have fixed this in 7+, then perhaps they could be back-ported to the macports GCC-6 branch.

However as I said in the Bugzilla - the upstream branch is closed (accepting no more changes) so that any fixes for this would have to be local.

**Comment by barracuda156** [comment]
> if you want to try and track down the changes that have fixed this in 7+, then perhaps they could be back-ported to the macports GCC-6 branch.
> 
> However as I said in the Bugzilla - the upstream branch is closed (accepting no more changes) so that any fixes for this would have to be local.

Are we sure that it is fixed in gcc7+? If so, perhaps not much point indeed.
